python-primitiv/primitiv/_trainer.pyx

```diff
@@ -1,38 +1,37 @@
 from primitiv import trainers as T
+from primitiv._parameter cimport _Parameter
+
 
 cdef class _Trainer:
 
-    @staticmethod
-    def load(str path):
-        name = _Trainer.detect_name(path);
-        if name == "SGD":
-            trainer = T.SGD()
-        elif name == "MomentumSGD":
-            trainer = T.MomentumSGD()
-        elif name == "AdaGrad":
-            trainer = T.AdaGrad()
-        elif name == "RMSProp":
-            trainer = T.RMSProp()
-        elif name == "AdaDelta":
-            trainer = T.AdaDelta()
-        elif name == "Adam":
-            trainer = T.Adam()
-        else:
-            raise OSError("Unknown trainer name: %s" % name)
-        trainer.set_configs_by_file(path)
-        return trainer
-
-    @staticmethod
-    def detect_name(str path):
-        return CppTrainer_detect_name(path.encode("utf-8")).decode("utf-8")
+    # NOTE(vbkaisetsu):
+    # This method should be called in the __init__() method of a
+    # custom Trainer class.
+    #
+    # Users can define custom trainers written in Python. This method
+    # generates an instance of a helper trainer called "PyTrainer" that
+    # can call methods implemented in child classes of Trainer.
+    def __init__(self):
+        if self.wrapped is not NULL:
+            raise MemoryError()
+        self.wrapped = new CppPyTrainer(self)
+
+    # NOTE(vbkaisetsu):
```

Comment:
This comment looks redundant because invoking `__dealloc__` recursively is a *specification* of Cython. Comments should include why this implementation exists.

Reply:
updated